include/zephyr/drivers/mspi.h

```diff
@@ -0,0 +1,791 @@
+	MSPI_DEVICE_CONFIG_NONE         = 0,
+	MSPI_DEVICE_CONFIG_CE_NUM       = BIT(0),
+	MSPI_DEVICE_CONFIG_FREQUENCY    = BIT(1),
+	MSPI_DEVICE_CONFIG_IO_MODE      = BIT(2),
+	MSPI_DEVICE_CONFIG_DATA_RATE    = BIT(3),
+	MSPI_DEVICE_CONFIG_CPP          = BIT(4),
+	MSPI_DEVICE_CONFIG_ENDIAN       = BIT(5),
+	MSPI_DEVICE_CONFIG_CE_POL       = BIT(6),
+	MSPI_DEVICE_CONFIG_DQS          = BIT(7),
+	MSPI_DEVICE_CONFIG_RX_DUMMY     = BIT(8),
+	MSPI_DEVICE_CONFIG_TX_DUMMY     = BIT(9),
+	MSPI_DEVICE_CONFIG_READ_INSTR   = BIT(10),
+	MSPI_DEVICE_CONFIG_WRITE_INSTR  = BIT(11),
+	MSPI_DEVICE_CONFIG_INSTR_LEN    = BIT(12),
+	MSPI_DEVICE_CONFIG_ADDR_LEN     = BIT(13),
+	MSPI_DEVICE_CONFIG_MEM_BOUND    = BIT(14),
+	MSPI_DEVICE_CONFIG_BREAK_TIME   = BIT(15),
+	MSPI_DEVICE_CONFIG_ALL          = BIT_MASK(16),
+};
+
+/**
+ * @brief MSPI XIP access permissions
+ */
+enum mspi_xip_permit {
+	MSPI_XIP_READ_WRITE     = 0,
+	MSPI_XIP_READ_ONLY      = 1,
+};
+
+/**
+ * @brief MSPI Configure API
+ * @defgroup mspi_configure_api MSPI Configure API
+ * @{
+ */
+
+/**
+ * @brief Stub for timing parameter
+ */
+enum mspi_timing_param {
+	MSPI_TIMING_PARAM_DUMMY
+};
+
+/**
+ * @brief Stub for struct timing_cfg
+ */
+struct mspi_timing_cfg {
+
+};
+
+/**
+ * @brief MSPI device ID
+ * The controller can identify its devices and determine whether the access is
+ * allowed in a multiple device scheme.
+ */
+struct mspi_dev_id {
+	/** @brief device gpio ce */
+	struct gpio_dt_spec     ce;
+	/** @brief device index on DT */
+	uint16_t                dev_idx;
+};
+
+/**
+ * @brief MSPI controller configuration
+ */
+struct mspi_cfg {
+	/** @brief mspi channel number */
+	uint8_t                 ui8MSPIChannel;
+	/** @brief Configure operaton mode */
+	enum mspi_op_mode       eOPMode;
+	/** @brief Configure duplex mode */
+	enum mspi_duplex        eDuplex;
+	/** @brief DQS support flag */
+	bool                    bDQS;
+	/** @brief Software managed multi-slave enable */
+	bool                    bSwMultiSlave;
+	/** @brief GPIO chip-select line (optional) */
+	struct gpio_dt_spec     *pCE;
+	/** @brief GPIO chip-select line numbers (optional) */
+	uint32_t                ui32CeGpioNum;
+	/** @brief Slave number from 0 to host controller slave limit. */
+	uint32_t                ui32SlaveNum;
+	/** @brief Maximum supported frequency in MHz */
+	uint32_t                ui32MaxFreq;
+	/** @brief Whether to re-initialize controller */
+	bool                    bReinit;
+};
+
+/**
+ * @brief MSPI DT information
+ */
+struct mspi_dt_spec {
+	/** @brief MSPI bus */
+	const struct device     *bus;
+	/** @brief MSPI hardware specific configuration */
+	struct mspi_cfg         config;
+};
+
+/**
+ * @brief MSPI controller device specific configuration
+ */
+struct mspi_dev_cfg {
+	/** @brief Configure CE0 or CE1 */
+	uint32_t                ui32CENum;
+	/** @brief Configure frequency */
+	uint32_t                ui32Freq;
+	/** @brief Configure I/O mode */
+	enum mspi_io_mode       eIOMode;
```

Comment:
I'm not sure I understand the idea of the IO mode parameter- many devices have some commands that support quad pad, while other commands only support single pad I/O. This feels like something we should be configuring at the transfer level?

Reply:
The settings in `mspi_dev_config` typically represent the most frequently used configurations required by the device. I know that some devices does not require explicit op-mode as a device current state but distinguishes the line modes only commands. But in real world usage, only one specific op-mode would be used for most of transfer operations for optimal performance. So it is not so different with devices that enters and exits certain op-mode by command. 

The reason why I split the configurations into mspi_dev_config and mspi_xfer is because switching between op-mode may require quite some time for both device and controller. To optimize performance, mspi_dev_config is dedicated for configuring the controller based on device requirements. Only a minimum device specific settings that may vary per transfer is included in mspi_xfer.

Reply:
My concern here is that the IO mode for read might be different than the IO mode for write. Perhaps the controller supports 1-8-8 read mode at the selected frequency, but only 1-4-4 write mode. I'm not familiar with the Ambiq controller- is there significant overhead to switch operating modes when using it? If so, how would something like the hypothetical device I describe above function?

Reply:
I'm aware that these type of device do exist, however not in the way you described but does involves ddr/sdr switching. I'm also considering combining the `mspi_op_mode` and `mspi_data_rate`.
These types of device should call `mspi_dev_config` for op mode switching and then call `mspi_transceive`.

The io mode switching in general should just be writing to controller registers. However, there are a number of factors we need to consider.
1. Does the device require mode switching also? If it does, would it require some time?
2. Does the controller have a hardware command queue, does it support changing all these settings and does the controller layer have access to it through HAL? 
   - If all of the above is true for everybody, then yes we can add it `mspi_xfer`. But I don't think this is the case.
   - If without a hardware command queue or without complete access to it, accessing controller configuration would be slower and you may have to configure the hardware every time when calling `mspi_transceive`. This is also what I really dislike the current SPI.  As we have so many layers of HAL API calls and Zephyr API calls. The hardware performance would get reduced very quickly.
3. The MSPI API is a lower API in the hierarchies of all Zephyr APIs. Thus having clear boundaries of each MSPI API should do is always better.
   - There is RTIO that act as a general work queue API people could use for chaining different MSPI API calls if the device operation mode switching become too tedious to handle.